include/random_walks/billiard_shake_and_bake_walk.hpp

```diff
@@ -0,0 +1,303 @@
+        using MT = typename Polytope::MT;
+        typedef typename Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> DenseMT;
+        static constexpr bool SPARSE = std::is_same_v<MT, Eigen::SparseMatrix<NT, Eigen::RowMajor>>;
+        using AA_type = std::conditional_t< SPARSE, typename Eigen::SparseMatrix<NT>, DenseMT >;
+
+        static constexpr NT kDefaultEpsilon = NT(1e-10);
+
+        template <typename GenericPolytope>
+        Walk(GenericPolytope &P, Point const& p, RandomNumberGenerator &rng,int facet_idx, unsigned int rho, NT eps = kDefaultEpsilon): P_{P}, epsilon_{eps},_rho{ (rho == 0) ? 1u : rho }
+        {
+            if(!P.is_normalized()) {
+                P.normalize();
+            }
+            _update_parameters = update_parameters();
+            std::pair<Point, NT> inner_ball = P.ComputeInnerBall();
+            _L = 2.0 * inner_ball.second;
+            if constexpr (SPARSE) {
+                _AA = (P.get_mat() * P.get_mat().transpose());
+            } else {
+                _AA.noalias() = (DenseMT)(P.get_mat() * P.get_mat().transpose());
+            }
+            initialize(p, facet_idx, rng);
+
+        }
+
+        template <typename GenericPolytope>
+        Walk(GenericPolytope &P, Point const& p, RandomNumberGenerator &rng, int facet_idx,    
+             parameters const& params, unsigned int rho, NT eps = kDefaultEpsilon): P_{P}, epsilon_{eps},_rho{ (rho == 0) ? 1u : rho }
+        {
+            if(!P_.is_normalized()) {
+                P_.normalize();
+            }
+            _update_parameters = update_parameters();
+            std::pair<Point, NT> inner_ball = P.ComputeInnerBall();
+            _L = 2.0 * inner_ball.second;
+            if constexpr (SPARSE) {
+                _AA = (P_.get_mat() * P_.get_mat().transpose());
+            } else {
+                _AA.noalias() = (DenseMT)(P_.get_mat() * P_.get_mat().transpose());
+            }
+            initialize(p, facet_idx, rng);
+
+        }
+
+        NT   get_epsilon() const noexcept { return epsilon_; }
+
+        void apply(unsigned int walk_len, RandomNumberGenerator& rng)
+        {
+            unsigned int n = P_.dimension();
+            typename Point::Coeff b;
+            NT* b_data;
+            if constexpr (SPARSE) {
+                b = P_.get_vec();
+                b_data = b.data();
+            }
+ 
+            for (auto j=0u; j<walk_len; ++j)
+            {
+                unsigned int rho_step =(_rho == 1)? 1    : 1 + static_cast<unsigned int>( rng.sample_urdist() * _rho );
+
+                NT T = -std::log(rng.sample_urdist()) * _L;
+                _v = get_direction (rng);
+                 _distances_set = BoundaryOracleHeap<NT>(P_.num_of_hyperplanes());
+
+                int it = 0;
+                std::pair<NT, int> pbpair = P_.line_first_positive_intersect(_p, _v, Ar_, Av_, _update_parameters);
+
+                Point p0 = _p;
+
+                if (T <= pbpair.first) {
```

Comment:
this check should become `if (k == 1) continue;` where `k` is the number of reflections (draw uniformly at random).

Reply:
similarly for the condition with `T` in the while loop